view/src/main/java/org/xcolab/view/pages/profile/beans/BadgeBean.java

```diff
@@ -1,82 +1,69 @@
 
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
 import org.xcolab.client.contest.ContestClientUtil;
-import org.xcolab.client.contest.exceptions.ContestNotFoundException;
-import org.xcolab.client.contest.pojo.Contest;
 import org.xcolab.client.contest.pojo.phases.ContestPhase;
 import org.xcolab.client.contest.pojo.phases.ContestPhaseRibbonType;
-import org.xcolab.client.proposals.ProposalAttributeClientUtil;
 import org.xcolab.client.proposals.ProposalClientUtil;
 import org.xcolab.client.proposals.ProposalPhaseClientUtil;
-import org.xcolab.client.proposals.enums.ProposalAttributeKeys;
 import org.xcolab.client.proposals.pojo.Proposal;
 import org.xcolab.client.proposals.pojo.phases.ProposalContestPhaseAttribute;
 import org.xcolab.util.enums.contest.ProposalContestPhaseAttributeKeys;
 import org.xcolab.view.pages.profile.entity.Badge;
 
 import java.io.Serializable;
 import java.util.ArrayList;
-import java.util.Date;
+import java.util.Comparator;
 import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
 
 public class BadgeBean implements Serializable {
 
     private static final long serialVersionUID = 1L;
-    private final static Logger _log = LoggerFactory.getLogger(BadgeBean.class);
-    private final long userId;
-    private final List<Badge> badges;
+
+    private final List<Badge> badges = new ArrayList<>();
 
     public BadgeBean(long userId) {
-        this.userId = userId;
-        badges = new ArrayList<>();
-        fetchBadges();
+        populateBadges(userId);
     }
 
-    private void fetchBadges() {
-        for(Proposal p : ProposalClientUtil.getMemberProposals(userId)) {
-            try {
-                final ContestPhaseRibbonType ribbonType = getRibbonType(p);
+    private void populateBadges(long userId) {
+        for (Proposal proposal : ProposalClientUtil.getMemberProposals(userId)) {
+            getRibbonAttribute(proposal)
+                    .ifPresent(ribbonAttribute -> {
+                        final ContestPhaseRibbonType ribbonType = ContestClientUtil
+                                .getContestPhaseRibbonType(ribbonAttribute.getNumericValue());
 
-                if (ribbonType != null  && ribbonType.getRibbon() > 0) {
-                    Contest contest = ProposalClientUtil.getCurrentContestForProposal(p.getProposalId());
-                    String proposalTitle = ProposalAttributeClientUtil
-                            .getProposalAttribute(p.getProposalId(), ProposalAttributeKeys.NAME, 0L)
-                            .getStringValue();
-                    badges.add(new Badge(ribbonType, p, proposalTitle, contest));
-                }
-            } catch (ContestNotFoundException e) {
-                _log.warn("Could not add badge to user profile view for userId: {} and proposalId: {}",
-                        p.getProposalId(), e);
-            }
+                        if (ribbonType.getRibbon() > 0) {
```

Comment:
I understand you didn't introduce this check, but to me as a reader that is kind of confusing. What exactly does it mean that `ribbonType.getRibbon()` is greater than 0?

Reply:
Yea, I wanted to remove this but unfortunately there is a ribbon type "0" in the database. I'm not quite sure what it means either. Ideally, I'd like to get rid of the possibility for it to be 0. I'll create a separate issue for that.